src/main/java/org/hibernate/build/gradle/inject/InjectionTask.java

```diff
@@ -43,64 +46,62 @@
 import javassist.bytecode.ClassFile;
 import javassist.bytecode.ConstantAttribute;
 import javassist.bytecode.FieldInfo;
-import org.gradle.api.Action;
-import org.gradle.api.Project;
-import org.gradle.api.Task;
-import org.gradle.api.tasks.SourceSet;
-import org.gradle.api.tasks.SourceSetContainer;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.TaskAction;
 
 /**
- * Responsible for coordinating the actual injection of project version.  Runs as a Gradle Action (doLast typically
- * to the main CompileJava task)
+ * Responsible for coordinating the actual injection of project version.
  *
  * @author Steve Ebersole
  */
-public class InjectionAction implements Action<Task> {
-	private static final Logger log = LoggerFactory.getLogger( InjectionAction.class );
+public abstract class InjectionTask extends DefaultTask {
+	static final String TASK_NAME = "injectVersion";
+	private static final Logger log = LoggerFactory.getLogger( InjectionTask.class );
 
 	private LoaderClassPath loaderClassPath;
 	private ClassPool classPool;
 
-	private final InjectionSpec injectionSpec;
+	@InputFiles
+	abstract ConfigurableFileCollection getClasspath();
 
-	public InjectionAction(InjectionSpec injectionSpec) {
-		this.injectionSpec = injectionSpec;
-	}
+	@Input
+	public abstract Property<String> getVersion();
+
+	@Input
+	public abstract ListProperty<TargetMember> getInjectionTargets();
```

Comment:
BTW, I've tried adding an output, as:

```java
	@OutputDirectory
	public FileCollection getOutputFiles() {
		return getClasspath();
	}
```

^ my thinking was since we update classes, it somewhat makes sense ... 

but I had build issues on the ORM side when I tried building an entire project, rather than just a single module where the injection plugin is actually applied. Some of those modules that didn't need the plugin were complaining... 

maybe it's because of that "global" declaration here: 

https://github.com/hibernate/hibernate-orm/blob/0080e17b43515f4a6e12218dd16dc879659e64a0/build.gradle#L22

🤔